documentation/source/extensions.rst

```diff
@@ -0,0 +1,111 @@
+
+- the package is in the ``cocotbext.EXTNAME`` namespace, and
+- the distribution (package) name is prefixed with ``cocotbext-EXTNAME``.
+
+Example:
+An SPI bus extension might be packaged as ``cocotbext-spi``, and its functionality would live in the ``cocotbext.spi`` namespace.
+The module can then be installed with ``pip3 install cocotbext-spi``, and used with ``import cocotbext.spi``.
+
+Types of cocotb extensions
+--------------------------
+
+For some types of cocotb extensions we have developed conventions which go beyond naming.
+These conventions help to achieve a consistent behavior across extensions of the same type.
+
+Bus extensions
+~~~~~~~~~~~~~~
+
+A cocotb extension which interacts with a bus or an interface (such as SPI or AXI) should build on top of a common set of classes to provide a uniform interface for its users.
+Typically, a bus extension provides three pieces of functionality:
+an abstraction of the bus itself, a bus driver, and a bus monitor.
+
+A bus driver is the "active" part, it drives the signals that make up the bus to request reads or writes from the bus.
+A bus monitor is the "passive" part, it observes signal changes on the bus and assigns meaning to them.
+Monitors can also check the bus behavior against a standard to ensure no invalid states are being observed.
+
+The signals which make up the bus should be grouped in a class inheriting from :any:`cocotb.bus.Bus`.
+Bus drivers should inherit from the :any:`cocotb.drivers.BusDriver` class.
+Bus monitors should inherit from the :any:`cocotb.monitors.BusMonitor` class.
+
+Packaging extensions
+--------------------
+
+To package a cocotb extension as Python package follow the :ref:`naming conventions <Naming conventions>`, and the `normal Python packaging rules <https://packaging.python.org/tutorials/packaging-projects/>`_.
+Extensions namespaced packages, implemented using the `native namespacing <https://packaging.python.org/guides/packaging-namespace-packages/#native-namespace-packages>`_ approach discussed in :pep:`420`.
+The module file hierarchy should be as follows (replace ``EXTNAME`` with the name of the extension, e.g. ``spi``).
+
+.. code-block::
+
+  # file structure of the cocotbext-EXTNAME repository
+  ├── cocotbext/
+  │   │   # No __init__.py here.
+  │   └── EXTNAME/
+  │       └── __init__.py
+  ├── README.md
+  └── setup.py
+
+The Python source code should go into the ``EXTNAME`` directory, next to the ``__init__.py`` file.
+All packaging metadata goes into ``setup.py``.
+
+.. code-block:: python3
+
+  # Minimal setup.py. Extend as needed.
+  from setuptools import setup, find_namespace_packages
+
+  setup(name = 'cocotbext-EXTNAME',
+        version = '0.1',
+        packages = find_namespace_packages(include=['cocotbext.*']),
+        install_requires = ['cocotb'],
+        python_requires = '>=3.5',
+        classifiers = [
+          "Programming Language :: Python :: 3",
+          "License :: OSI Approved :: MIT License",
```

Comment:
Do we want to express a requirement or (strong) recommendation that the extension be under the MIT License?

Reply:
Not really. We'll most likely have requirements and recommendations for code we host in the cocotb org, and for things we endorse, but that's not yet discussed and shouldn't be in this document. I'll remove this classifier.

Reply:
I'd lean towards leaving this as is, but add a comment saying `# or another license of your choice`. Otherwise we risk a bunch of extensions that people forget to license that no one can actually use.

Reply:
The license would also need to go into the `license` field and a LICENSE File (or similar) is also needed. I opened https://github.com/cocotb/cocotb/issues/1362 to gather thoughts on recommended licenses (this discussion is probably larger than this small comment sub-thread), and I'd go for adding this information in a follow-up as soon as we have consensus in the issue.